### DISCUSSION TITLE
Compiler enforced pureness by default for `fn`

### DISCUSSION THREAD

**Discussion by FelixBenning** [modular/modular#345] [Ideas]
It might make sense to have `fn` be pure by default (compiler enforced) and require an `impure` annotation to allow impure behavior. It might even make sense for `fn` to be the pure function type and `def` to allow side effects, but it is definitely worth considering this at the start since defaults influence peoples behavior. Similar to how newer languages make variabels immuatable by default and require a `mutable` annotation in some form.

### Motivation

I am not a compiler engineer, but I would assume that optimizations can be made for pure functions. Even if the compiler can detect pure functions without the keyword, the keyword would be helpful for the user to ensure that they did not accidentally make the function impure costing them performance. Some libraries like jax currently require functions to be pure but there is no way to enforce this pureness in python right now so you have to pray you are not making mistakes. This is similar to informal interfaces without static checking (features like `@abstractmethod`), which result in many bugs.

If this was the default behaviour from the start, then more functions would be written like this and a greater share of the library would benefit from optimizations for pure functions. People would start out writing a pure function and only add the `impure` keyword as needed because the compiler complains or they realize they can not avoid it. The introduction of a new type of function is a good time to consider this I think.

If you decide that you would rather have `fn` allow side-effects by default, the introduction of a `pure` keyword is likely akin to the introduciton of `public/private` keywords which you explicitly want to postpone. So this discussion should focus on the question whether it makes sense to have this behavior by default. I honestly don't know, just wanted to put this thought out before you reach 1.0 and can't change your mind anymore.

**Comment by lattner** [comment]
We can't get away with this I'm afraid.  We want people to be able to mix and match fn and def (therefore only affecting the internal implementation details) and def's would have to be treated as side effecting.  Going with this proposal would mean that fn's couldn't call defs, or (if there was an exception) "being pure" couldn't be counted to mean anything

**Reply by FelixBenning** [reply]
I guess that only leaves the version where you can still call def if you annotate `fn` with `impure`.

Besides the nudging effect, this would probably also reflect the way people will usually use def and fn. fn will probably be used if you want to be closer to the metal with more performance guarantees (i.e. in libraries), def will probably be more glue code. So I think usually def will call fn. So the annotation would likely be the exception not the rule.

**Comment by lattner** [comment]
Pureness is what is known as an "effect" in PL terminology.  You can see this in the handling of async and raises in the current mojo implementation: a non-raising function is not allowed to call a raising function directly - it must wrap it in a try block.

I don't see a way to provide this sort of mapping from one world to the other for purity, I think we cannot practically implement this, and while pure computation is important, it is actually quite compilicated: is reading from memory pure?  If no, "purity" is pretty useless.  If so, you cannot use purity information for much optimization, because you need to know which memory sets may be read and written by functions anyway.

Also, in other pure-functional languages like haskell, you need escape hatches (perform unsafe io) because you want to add printf debugging etc to "pure" functions and compiler enforcement makes that whole thing incredibly difficult.

overall I can understand wanting to have this conceptually, but I can't see how it could work out well in practice.  We can come back to this later as the language evolves.

**Reply by FelixBenning** [reply]
So my definition of pureness would be two things:
1. If you pass in the same* input you are going to get the same output
2. The function has noch side-effects (i.e. you could cache inputs and return your output cache instead of calling the function again with the same inputs)

*What you are probably getting at with the "is reading from memory pure?" question is: If I pass an array, I am passing a pointer. If I pass in the same pointer again later (but the array was changed) this would result in a different output so reading from memory would not be pure in that sense. But we would typically also say that these two arrays are no longer equal. And in this way this question is closely related to the `==` operator. We typically implement `==` in such a way, that the underlying memory of two arrays must be the same, not the pointer. So in this sense reading from memory is pure. On the other hand if we use the `==` on a `Path` object, then only the filename counts (and not its contents). So reading from a file is not pure.

That is how I would define it. There might be better definitions and this definition might be useless for optimization.
